### PR TITLE
CoolProp integration; Pressure-aware water property calculations (v3.4.2)

### PR DESCRIPTION
**Pull request by softwareengineerprogrammer**
Self-reviewed sub-PRs:
1. https://github.com/softwareengineerprogrammer/GEOPHIRES-X/pull/10
    1. Addresses https://github.com/NREL/GEOPHIRES-X/issues/113
1. https://github.com/softwareengineerprogrammer/GEOPHIRES-X/pull/11
    1. https://github.com/NREL/GEOPHIRES-X/issues/110
    1. Fixes bugs related to resevoir hydrostatic pressure, discovered in the course of implementation
    1. Also relevant to https://github.com/NREL/GEOPHIRES-X/issues/118
1. Addresses https://github.com/NREL/GEOPHIRES-X/issues/107
1. Identifies and partially addresses https://github.com/NREL/GEOPHIRES-X/issues/120

Reviewers are encouraged to look at diff to example outputs (https://github.com/NREL/GEOPHIRES-X/pull/121/files#diff-34b4448aead45f08341092fe535fc9270bcf1d8420c517)

---

Deployed and verified on backend/UI (https://jonathanpezzino.com/geothermal/geophires)

Screenshot 2024-02-14 at 16 32 15

### CODE REVIEW

File: tests/examples/Wanju_Yuan_Closed-Loop_Geothermal_Energy_Recovery.out [hunk 1]
```diff
@@ -4,21 +4,21 @@
...
                            ***SUMMARY OF RESULTS***
 
       End-Use Option: Electricity
       Average Net Electricity Production:                     1.10 MW
-      Electricity breakeven price:                           37.41 cents/kWh
+      Electricity breakeven price:                           44.08 cents/kWh
       Number of production wells:                             1
       Number of injection wells:                              0
       Flowrate per production well:                         110.0 kg/sec
-      Well depth (or total length, if not vertical):          9.0 kilometer
```
Comment (softwareengineerprogrammer) [inline tests/examples/Wanju_Yuan_Closed-Loop_Geothermal_Energy_Recovery.out:21]: Fixed incorrect CLGS calculation of total drilled length

File: tests/examples/example1.out [hunk 2]
```diff
@@ -66,9 +66,9 @@ Simulation Metadata
       Well seperation: fracture height:                     900.00 meter
       Fracture area:                                     810000.00 m**2
       Reservoir volume:                              1000000000 m**3
-      Reservoir hydrostatic pressure:                       100.00 kPa
```
Comment (softwareengineerprogrammer) [inline tests/examples/example1.out:69]: Note previous examples all had incorrect default value of 100.00 kPa, examples are now updated with calculated hydrostatic pressure, and default is now ~29k kPa

**Review by malcolm-dsider — APPROVED** [review]
Looks good, but I don't understand the new construct on Model.py:

self.reserv: Reservoir = TDPReservoir(self)

**Comment by softwareengineerprogrammer** [comment]
> Looks good, but I don't understand the new construct on Model.py:
> `self.reserv: Reservoir = TDPReservoir(self)`

The addition of `: Reservoir` is a type hint. In this case I added the hint because it enabled PyCharm to give me better autocomplete suggestions as I was working on code that referenced `model.reserv` (i.e. typing `model.reserv.` and then being presented with suggestions for `.value`, `.quantity()`, etc.).

In general, we should ideally be using/adding type hints for all code as best practice since they improve contributor experience, code comprehensibility, and serve as a guardrail against many common classes of programming errors. I have created a tracking item for this: https://github.com/NREL/GEOPHIRES-X/issues/100. However, I am deferring on adding types everywhere for a few reasons for now:

1. Adding type hints everywhere all at once would be a major task with limited immediate functional benefit and lots of risk for breakage/issues. The short-term ROI isn't worth it given current resourcing and priorities (although I look forward to the day/time that it will be)
2. The work would entail adding a type linter so that type hinting is automatically enforced instead of relying on contributors to remember on their own - similar to how `pre-commit` is enabled for styling. Setting up type linting (in addition to type hinting) adds significant work

So for now, I am just doing the following:
1. Adding type hints to new code that I add (at least when I remember - hence the value of the linter)
2. Adding type hints to existing code where there is immediate benefit relevant to the work I'm doing (such as enabling autocomplete that helped me implement this PR more efficiently)

File: src/geophires_x/AGSWellBores.py [hunk 3]
```diff
@@ -1078,7 +1102,10 @@ def Calculate(self, model: Model) -> None:
             self.ProducedTemperature.value = self.InterpolatedTemperatureArray.copy()
 
             tot_length, vert_length, horizontal_lengths = self.calculatedrillinglengths(model)
-            model.reserv.depth.value = model.reserv.InputDepth.value * 1000.0  # in this case, reserv.depth is just the vertical drill depth
+
+            # in this case, reserv.depth is just the vertical drill depth
+            # FIXME earlier calculations use depth before this value is set, meaning they're using the wrong value
```
Comment (softwareengineerprogrammer) [inline src/geophires_x/AGSWellBores.py:1107]: Created follow-up issue: https://github.com/NREL/GEOPHIRES-X/issues/122

**Comment by malcolm-dsider** [comment]
It seems logical, and I will endeavor to follow your suggestions below about when to add type hints… but I clearly don’t know how to add the proper type hints in all cases…

**Comment by malcolm-dsider** [comment]
This has been addressed by replacing all references to reserv.depth to reserv.InputDepth for the AGSWellBores object. reserve.InputDepth is the correct value to be using in the AGS case, where the InputDepth of the start of the multilateral section is different from the OutputDepth (multilaterals are not required to be horizontal). The references were created with a cut-and-paste error, I think.

 

This only affected the calculations in the case where we are calling Wnaju’s code (i.e. e, multilateral closed loop or single-lateral closed loop when temperatures are above 375C and the CLGS code will fail).  

 

The net result of this change was very small (<1%) in a few of the output report values, so I updated the output report with those small changes.

 

How would you like me to proceed with respect to GitHub for the changes I made related to this issue only? I have several changes waiting for commits, etc, related to other stuff I am doing, so the only changes I made to the AGSWellbore object were the corrections for this error.

File: src/geophires_x/AGSWellBores.py (hunk 3 above)
Comment (softwareengineerprogrammer) [reply src/geophires_x/AGSWellBores.py:1107]: Moved comment reply from email to https://github.com/NREL/GEOPHIRES-X/issues/122#issuecomment-1948693134

**Pull request merged**